### PR TITLE
StyleGAN Model ADNI Dataset 47492347

### PR DESCRIPTION
**Pull request by jjntw**
Hello team, this is my working implementation of a StyleGAN generative model trained on the ADNI brain dataset using PyTorch. Please have a read through my README file for a detailed description of my implementation details.

Summary of files:
<ul>
<li>dataset.py ~ contains data loading preprocessing functionalities</li>
<li>embedding.py ~ functions used to plot UMAP embeddings</li>
<li>modules.py ~ collection of modules and models that make up StyleGAN architecture</li>
<li>predict.py ~ test script used for generating images using a pre-trained model</li>
<li>settings.py ~ config file used for tweaking train hyperparameters</li>
<li>train.py ~ main model training script</li>
<li>utils.py ~ contains global imports</li>
</ul>

Feel free to inquire about any issues to fix or potential improvements!
Jun Woo (Jason) Jung
SN: 47492347

### CODE REVIEW

**Comment by Uto0op** [comment]
Recognition problem: StyleGAN on ADNI ---- Hard Difficulty

1. Algorithm solves the problem: (4 Marks)
Clear generated images;
Losses of the generator are unusual but left unexplained -0.5;
UMAP plotted on training data instead of the style space -0.5

2. Algorithm demonstrated and functions: (3 Marks)

4. Design: (0.5 Mark)
Not very modularized design -0.5

5. Commenting:
Sufficient docstrings and comments (1 Mark)

6. Algorithm above Normal Difficulty: (5 Marks)

7. Algorithm is Hard difficulty: (5 Marks)

**Suggestions**:
- I can see you intend to achieve modularization, the problem is, however, only 5 scripts need to be uploaded (including utils.py, see Section 4 on task sheet). In your code, other scripts are not really necessary, e.g. functions in `embedding.py` should either go to `dataset.py` or `utiles.py`, so are the parameters in `settings.py`;
- Embedding should be plotted on the style space according to the reference paper, and by doing so hopefully you can see it makes more sense;
- The losses, especially ones from the generator, are abnormal --- this is fine since GANs are tricky. But insteading of displaying the loss plots, you should also attempt to justify how come the losses.

**Comment by yexincheng** [comment]
**This is an initial inspection, no action is required at this point**
1. **Solves problem**: Nice documentation and clear explanation of the network. The generated images are reasonable and have clear structures. Reasonable loss plots and nice records of all experiment steps. UMAP does not show clear clusters, but it is a good attempt to explore the reason.
2. **Implementation**: Good usage of ArgParse. Settings could be args in the training script. Overall nice
3. **Design**: Great
4. **Comment**: Sufficient and well-structured.
5. **Difficulty**: Hard

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Adequate design and implementation. </td><td>-1</td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>Header blocks. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>OK solution to problem. Loss unexplianed.</td><td>-1</td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td>-1</td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard. StyleGAN (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Some/Adequate Progressive commits. Somewhat minimal.</td><td>-1</td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF Checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). ON time.</td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-4</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by shakes76** [comment]
Good work!
